src/Container.php

```diff
@@ -31,11 +31,11 @@ final class Container
     /**
      * Static instance lookup.
      *
-     * @return static
+     * @return self
      */
     public static function instance()
     {
-        if (!self::$instance) {
+        if (! self::$instance instanceof self) {
```

Comment:
I understand this makes phpstan happy, but in PHP does not exists an empty/false object, any objects is casted to true, even an empty `stdClass`. So if a type is `null|object` checking for its boolean value is perfectly fine. This makes it more verbose (and at micro level possibly less performant) for no reason. Actually, enforce this opinionated code style is the main reason I don't like php stan: I want static analisys to find bugs, not find code style issue.

Reply:
putting a non-boolean value in if() stop you from writing robust code

Reply:
> in PHP does not exists an empty/false object, any objects is casted to true

Not necessarily true.... See: https://phpcheatsheets.com/test/boolean_tests/ (check the last two columns and the last few rows)

Reply:
> Not necessarily true...
The only case in which an object implicitly cast to boolean returns false is the object obtained via `SplBool(false)`. And `SplBool` is a class that comes from an extension not bundled with PHP (see https://www.php.net/manual/en/spl-types.installation.php). Meaning, that if we consider only classes in PHP _core_, the statement is always true.

> putting a non-boolean value in if() stops you from writing robust code
@szepeviktor that's very opinionated, and I kindly disagree.
As said above, if you are sure that a variable is `ClassName|null` (and that `ClassName` is not `SplBool`, to make @jrfnl happy :)), then `if ($var)` is identical to if ($var instanceof ClassName). And saying that two statements with 100% identical results have different "robustness" is nonsensical to me: they can be more/less readable, they can be more/less performant, but they can't be less/more robust, as the robustness depends on how they behave on different inputs and conditions, and if they always behave the same, I don't see how they can be differently robust.
Moreover, in PHP implicit casting has its rules that you can leverage.
For example if a variable is of type `array|null` then `if ($var)` is equivalent to `if ($var !== null && $var !== [])`, and if that is what you want to test for, then `if ($var)` is much more readable and much less error prone (it's easy to mistype `!=` for `!==`).
So in that specific case, IMO, not putting a boolean in `if` is much better than doing it.
In short, the implicit boolean casting of variables is a _tool_ in PHP. And just like any tool might be used for good.